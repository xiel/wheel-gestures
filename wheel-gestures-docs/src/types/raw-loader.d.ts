declare module '!!raw-loader!*' {
  const content: string
  export = content
}
