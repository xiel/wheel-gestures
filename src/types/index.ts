/**
 * public types
 * these types are exported by the module
 */
export type VectorXYZ = [number, number, number]
export type BooleanXYZ = [boolean, boolean, boolean]
export type PreventWheelActionType = true | 'x' | 'y' | 'z' | false
export type ReverseSign = BooleanXYZ | boolean

export interface WheelGesturesConfig {
  preventWheelAction: PreventWheelActionType
  reverseSign: ReverseSign
}

// all options are optional and have reasonable defaults
export type WheelGesturesOptions = Partial<WheelGesturesConfig>
export type WheelEventDataRequiredFields = 'deltaMode' | 'deltaX' | 'deltaY' | 'timeStamp'

export interface WheelEventData
  extends Pick<WheelEvent, WheelEventDataRequiredFields>,
    Partial<Omit<WheelEvent, WheelEventDataRequiredFields>> {}

export interface WheelEventState {
  isStart: boolean
  isMomentum: boolean
  isEnding: boolean
  isMomentumCancel: boolean
  axisMovement: VectorXYZ
  axisVelocity: VectorXYZ
  axisDelta: VectorXYZ
  event: WheelEvent | WheelEventData
  previous?: WheelEventState
}

export type WheelGesturesEventMap = {
  wheel: WheelEventState
}
