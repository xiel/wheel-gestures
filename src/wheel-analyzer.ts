import {
  Axis,
  DeltaProp,
  PreventWheelActionType,
  ScrollPoint,
  SubscribeFn,
  Unobserve,
  Unsubscribe,
  WheelEventData,
  WheelPhase,
} from './wheel-analyzer.types'
import { normalizeWheel } from './normalizer/wheel-normalizer'

const WHEELEVENTS_TO_MERGE = 2
const WHEELEVENTS_TO_ANALAZE = 5

const SOON_ENDING_WHEEL_COUNT = 3
const SOON_ENDING_THRESHOLD = 1.4

const ACC_FACTOR_MIN = 0.6
const ACC_FACTOR_MAX = 0.96
const DELTA_MAX_ABS = 150

// the initial timeout period is pretty long, so even old mouses, which emit wheel events less often, can produce a continuous gesture
// the timeout is automatically adjusted during a gesture
const WILL_END_TIMEOUT_DEFAULT = 400

const axes: [Axis, Axis] = ['x', 'y']
const deltaProp: Record<Axis, DeltaProp> = {
  x: 'deltaX',
  y: 'deltaY',
}

export interface Options {
  preventWheelAction: PreventWheelActionType
  isDebug?: boolean
}

const defaults: Options = {
  preventWheelAction: 'all',
  isDebug: process.env.NODE_ENV === 'development',
}

export class WheelAnalyzer {
  private isStarted = false
  private isStartPublished = false
  private isMomentum = false
  private lastAbsDelta = Infinity
  private axisDeltas: number[] = [0, 0]
  private axisVelocity: number[] = [0, 0]
  private accelerationFactors: number[][] = []
  private deltaVelocity = 0 // px per second
  private deltaTotal = 0 // moved during this scroll interaction

  private scrollPoints: ScrollPoint[] = []
  private scrollPointsToMerge: ScrollPoint[] = []

  private subscriptions: SubscribeFn[] = []
  private targets: EventTarget[] = []

  private options: Options
  private willEndTimeout = WILL_END_TIMEOUT_DEFAULT

  private willEnd = (() => {
    let willEndId = setTimeout(() => {}, this.willEndTimeout)
    return () => {
      clearTimeout(willEndId)
      willEndId = setTimeout(this.end, this.willEndTimeout)
    }
  })()

  public constructor(options: Partial<Options> = {}) {
    // merge passed options with defaults (filter undefined option values)
    this.options = Object.entries(options)
      .filter(([, value]) => value !== undefined)
      .reduce((o, [key, value]) => Object.assign(o, { [key]: value }), { ...defaults })
  }

  public observe = (target: EventTarget): Unobserve => {
    target.addEventListener('wheel', this.feedWheel as EventListener, { passive: false })
    this.targets.push(target)
    return this.unobserve.bind(this, target)
  }

  public unobserve = (target: EventTarget) => {
    target.removeEventListener('wheel', this.feedWheel as EventListener)
    this.targets = this.targets.filter((t) => t !== target)
  }

  // stops watching all of its target elements for visibility changes.
  public disconnect = () => {
    this.targets.forEach(this.unobserve)
  }

  public subscribe = (callback: SubscribeFn): Unsubscribe => {
    this.subscriptions.push(callback)
    // return bound unsubscribe
    return this.unsubscribe.bind(this, callback)
  }

  public unsubscribe = (callback: SubscribeFn) => {
    if (!callback) throw new Error('please pass the callback which was used to subscribe')
    this.subscriptions = this.subscriptions.filter((s) => s !== callback)
  }

  public feedWheel = (wheelEvents: WheelEventData | WheelEventData[]) => {
    if (!wheelEvents) return

    if (Array.isArray(wheelEvents)) {
      wheelEvents.forEach((wheelEvent) => this.processWheelEventData(wheelEvent))
    } else {
      this.processWheelEventData(wheelEvents)
    }
  }

  private publish = (phase: WheelPhase, data = this.getCurrentState(phase)) => {
    this.subscriptions.forEach((fn) => fn(phase, data))
  }

  private shouldPreventDefault(e: WheelEventData) {
    const { preventWheelAction } = this.options
    const { deltaX, deltaY } = e

    switch (preventWheelAction) {
      case 'all':
        return true
      case 'x':
        return Math.abs(deltaX) >= Math.abs(deltaY)
      case 'y':
        return Math.abs(deltaY) >= Math.abs(deltaX)
    }

    this.debugMessage('unsupported preventWheelAction value: ' + preventWheelAction, 'warn')
  }

  private clampDelta(delta: number) {
    return Math.min(DELTA_MAX_ABS, Math.max(-DELTA_MAX_ABS, delta))
  }

  private processWheelEventData(wheelEvent: WheelEventData) {
    const normalizedWheel = normalizeWheel(wheelEvent)

    if (wheelEvent.preventDefault && this.shouldPreventDefault(wheelEvent)) {
      wheelEvent.preventDefault()
    }

    if (!this.isStarted) {
      this.start()
    }

    const currentDelta = this.clampDelta(
      Math.abs(normalizedWheel.deltaY) > Math.abs(normalizedWheel.deltaX)
        ? normalizedWheel.deltaY
        : normalizedWheel.deltaX
    )
    const currentAbsDelta = Math.abs(currentDelta)

    if (this.isMomentum && currentAbsDelta > this.lastAbsDelta) {
      this.end()
      this.start()
    }

    this.axisDeltas = this.axisDeltas.map(
      (prevDelta, i) => prevDelta + this.clampDelta(normalizedWheel[deltaProp[axes[i]]])
    )
    this.deltaTotal = this.deltaTotal + currentDelta
    this.lastAbsDelta = currentAbsDelta

    this.scrollPointsToMerge.push({
      currentDelta: currentDelta,
      currentAbsDelta: currentAbsDelta,
      axisDeltaUnclampt: [normalizedWheel.deltaX, normalizedWheel.deltaY],
      timestamp: wheelEvent.timeStamp || Date.now(),
    })

    if (this.scrollPointsToMerge.length === WHEELEVENTS_TO_MERGE) {
      const mergedScrollPoint: ScrollPoint = {
        currentDelta: this.scrollPointsToMerge.reduce((sum, b) => sum + b.currentDelta, 0) / WHEELEVENTS_TO_MERGE,
        currentAbsDelta: this.scrollPointsToMerge.reduce((sum, b) => sum + b.currentAbsDelta, 0) / WHEELEVENTS_TO_MERGE,
        axisDeltaUnclampt: this.scrollPointsToMerge.reduce(
          ([sumX, sumY], { axisDeltaUnclampt: [x, y] }) => [sumX + x, sumY + y],
          [0, 0]
        ),
        //.map((sum) => sum / WHEELEVENTS_TO_MERGE),
        timestamp: this.scrollPointsToMerge.reduce((sum, b) => sum + b.timestamp, 0) / WHEELEVENTS_TO_MERGE,
      }

      this.scrollPoints.push(mergedScrollPoint)

      // only update velocity after a merged scrollpoint was generated
      this.updateVelocity()

      if (!this.isMomentum) {
        this.detectMomentum()
      }

      // reset merge array
      this.scrollPointsToMerge = []
    }

    if (!this.scrollPoints.length) {
      this.updateStartVelocity()
    }

    // publish start after all data points have been updated
    // TODO: check momentum afterwards
    if (!this.isStartPublished) {
      this.publish(WheelPhase.ANY_WHEEL_START)
      this.publish(WheelPhase.WHEEL_START)
      this.isStartPublished = true
    }

    this.publish(WheelPhase.ANY_WHEEL)
    this.publish(this.isMomentum ? WheelPhase.MOMENTUM_WHEEL : WheelPhase.WHEEL)

    // calc debounced end function, to recognize end of wheel event stream
    this.willEnd()
  }

  private debugMessage(msg: string, level: keyof Console = 'log') {
    if (!this.options.isDebug) return
    console[level](msg)
  }

  private updateStartVelocity() {
    const latestScrollPoint = this.scrollPointsToMerge[this.scrollPointsToMerge.length - 1]
    this.axisVelocity = latestScrollPoint.axisDeltaUnclampt.map((d) => d / WILL_END_TIMEOUT_DEFAULT)
  }

  private updateVelocity() {
    // need to have two recent points to calc velocity
    const [pA, pB] = this.scrollPoints.slice(-2)

    if (!pA || !pB) {
      return
    }

    // time delta
    const deltaTime = pB.timestamp - pA.timestamp

    if (deltaTime <= 0) {
      this.debugMessage('invalid deltaTime')
      return
    }

    // calc the velocity per axes
    const velocity = pB.axisDeltaUnclampt.map((d) => d / deltaTime)

    // calc the acceleration factor per axis
    const accelerationFactor = velocity.map((v, i) => {
      return v / (this.axisVelocity[i] || 1)
    })

    this.axisVelocity = velocity
    this.accelerationFactors.push(accelerationFactor)
    this.updateWillEndTimeout(deltaTime)
  }

  private updateWillEndTimeout(deltaTime: number) {
    // use current time between events rounded up and increased by a bit as timeout
    let newTimeout = Math.ceil(deltaTime / 10) * 10 * 1.2

    // double the timeout, when momentum was not detected yet
    if (!this.isMomentum) {
      newTimeout = Math.max(100, newTimeout * 2)
    }

    this.willEndTimeout = Math.min(1000, Math.round(newTimeout))
  }

  private static accelerationFactorInMomentumRange(accFactor: number) {
    // when main axis is the the other one and there is no movement/change on the current one
    if (accFactor === 0) return true
    return accFactor <= ACC_FACTOR_MAX && accFactor >= ACC_FACTOR_MIN
  }

  private detectMomentum() {
    if (this.accelerationFactors.length < WHEELEVENTS_TO_ANALAZE) {
      return this.isMomentum
    }

    const recentAccelerationFactors = this.accelerationFactors.slice(WHEELEVENTS_TO_ANALAZE * -1)

    // check recent acceleration / deceleration factors
    const detectedMomentum = recentAccelerationFactors.reduce((mightBeMomentum, accFac) => {
      // all recent need to match, if any did not match -> short circuit
      if (!mightBeMomentum) return false
      // when both axis decelerate exactly in the same rate it is very likely caused by momentum
      const sameAccFac = !!accFac.reduce((f1, f2) => (f1 && f1 < 1 && f1 === f2 ? 1 : 0))
      // check if acceleration factor is within momentum range
      const bothAreInRangeOrZero =
        accFac.filter(WheelAnalyzer.accelerationFactorInMomentumRange).length === accFac.length
      // one the requirements must be fulfilled
      return sameAccFac || bothAreInRangeOrZero
    }, true)

    // only keep the most recent events
    this.accelerationFactors = recentAccelerationFactors

    if (detectedMomentum && !this.isMomentum) {
      this.publish(WheelPhase.WHEEL_END)
      this.isMomentum = true
      this.publish(WheelPhase.MOMENTUM_WHEEL_START)
    }

    return this.isMomentum
  }

  private getDebugState(): Partial<WheelAnalyzer> {
    const { ...props } = this
    return props
  }

  public getCurrentState(type: WheelPhase) {
    const debugData = this.options.isDebug ? this.getDebugState() : null
    return {
      type,
      debugData,
      willEndSoon: this.willEndSoon,
      isMomentum: this.isMomentum,
      deltaVelocity: this.deltaVelocity,
      deltaTotal: this.deltaTotal,
      axisDeltas: this.axisDeltas,
      axisVelocity: this.axisVelocity,
    }
  }

  private start() {
    this.isStarted = true
    this.isStartPublished = false
    this.isMomentum = false
    this.lastAbsDelta = Infinity
    this.axisDeltas = [0, 0]
    this.axisVelocity = [0, 0]
    this.deltaVelocity = 0
    this.deltaTotal = 0
    this.scrollPointsToMerge = []
    this.scrollPoints = []
    this.accelerationFactors = []
    this.willEndTimeout = WILL_END_TIMEOUT_DEFAULT
  }

  private end = () => {
    if (!this.isStarted) return

    if (this.isMomentum) {
      if (!this.willEndSoon) {
        this.publish(WheelPhase.MOMENTUM_WHEEL_CANCEL)
      } else {
        this.publish(WheelPhase.MOMENTUM_WHEEL_END)
      }
    } else {
      // in case of momentum, this event was already triggered when the momentum was detected so we do not trigger it here
      this.publish(WheelPhase.WHEEL_END)
    }

    this.publish(WheelPhase.ANY_WHEEL_END)

    this.isMomentum = false
    this.isStarted = false
  }

  private get willEndSoon() {
    const absDeltas = this.scrollPoints
      .slice(SOON_ENDING_WHEEL_COUNT * -1)
      .map(({ currentAbsDelta }) => currentAbsDelta)
    const absDeltaAverage = absDeltas.reduce((a, b) => a + b, 0) / absDeltas.length
    return absDeltaAverage <= SOON_ENDING_THRESHOLD
  }
}
