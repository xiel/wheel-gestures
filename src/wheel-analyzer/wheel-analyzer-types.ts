export type VectorXYZ = [number, number, number]
export type BooleanXYZ = [boolean, boolean, boolean]

export interface WheelAnalyzerState {
  isStarted: boolean
  isStartPublished: boolean
  isMomentum: boolean
  startTime: number
  lastAbsDelta: number
  axisMovement: VectorXYZ
  axisVelocity: VectorXYZ
  accelerationFactors: number[][]
  scrollPoints: MergedScrollPoint[]
  scrollPointsToMerge: ScrollPoint[]
  willEndTimeout: number
}

export interface PhaseData {
  type: WheelPhase
  isMomentum: boolean
  axisMovement: VectorXYZ
  axisVelocity: VectorXYZ
  axisDelta: VectorXYZ
  event: WheelEvent | WheelEventData
}

export interface ScrollPoint {
  deltaMaxAbs: number
  axisDelta: VectorXYZ
  timeStamp: number
}

export interface MergedScrollPoint {
  deltaMaxAbsAverage: number
  axisDeltaSum: VectorXYZ
  timeStamp: number
}

export type WheelEventDataRequiredFields = 'deltaMode' | 'deltaX' | 'deltaY' | 'timeStamp'

export interface WheelEventData
  extends Pick<WheelEvent, WheelEventDataRequiredFields>,
    Partial<Omit<WheelEvent, WheelEventDataRequiredFields>> {}

export type WheelTypes = 'WHEEL' | 'ANY_WHEEL' | 'MOMENTUM_WHEEL'

export enum WheelPhase {
  'ANY_WHEEL_START' = 'ANY_WHEEL_START',
  'ANY_WHEEL' = 'ANY_WHEEL',
  'ANY_WHEEL_END' = 'ANY_WHEEL_END',
  'WHEEL_START' = 'WHEEL_START',
  'WHEEL' = 'WHEEL',
  'WHEEL_END' = 'WHEEL_END',
  'MOMENTUM_WHEEL_START' = 'MOMENTUM_WHEEL_START',
  'MOMENTUM_WHEEL' = 'MOMENTUM_WHEEL',
  'MOMENTUM_WHEEL_CANCEL' = 'MOMENTUM_WHEEL_CANCEL',
  'MOMENTUM_WHEEL_END' = 'MOMENTUM_WHEEL_END',
}

export type SubscribeFn = (type: WheelPhase, data: PhaseData) => void
export type Unsubscribe = () => void

export type Unobserve = () => void
export type PreventWheelActionType = 'all' | 'x' | 'y'
export type ReverseSign = BooleanXYZ | boolean
