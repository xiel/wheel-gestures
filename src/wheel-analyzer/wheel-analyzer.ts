import { average } from '../utils/utils'
import { normalizeWheel } from '../wheel-normalizer/wheel-normalizer'
import { createWheelAnalyzerState } from './state'
import {
  Axis,
  DeltaProp,
  PhaseData,
  PreventWheelActionType,
  ScrollPoint,
  SubscribeFn,
  Unobserve,
  Unsubscribe,
  WheelEventData,
  WheelPhase,
} from './wheel-analyzer-types'

const SOON_ENDING_THRESHOLD = 1.4
const ACC_FACTOR_MIN = 0.6
const ACC_FACTOR_MAX = 0.96
const DELTA_MAX_ABS = 150

const axes: [Axis, Axis] = ['x', 'y']
const deltaProp: Record<Axis, DeltaProp> = {
  x: 'deltaX',
  y: 'deltaY',
}

const isDev = process.env.NODE_ENV !== 'production'
const WHEELEVENTS_TO_MERGE = 2
const WHEELEVENTS_TO_ANALAZE = 5
const SOON_ENDING_WHEEL_COUNT = 3

export interface Options {
  preventWheelAction: PreventWheelActionType
}

const defaults: Options = {
  preventWheelAction: 'all',
}

export function WheelAnalyzer(optionsParam: Partial<Options> = {}) {
  let state = createWheelAnalyzerState()
  let subscriptions: SubscribeFn[] = []
  let targets: EventTarget[] = []
  let currentEvent: WheelEventData

  // merge passed options with defaults (filter undefined option values)
  const options: Options = Object.entries(optionsParam)
    .filter(([, value]) => value !== undefined)
    .reduce((o, [key, value]) => Object.assign(o, { [key]: value }), { ...defaults })

  const observe = (target: EventTarget): Unobserve => {
    target.addEventListener('wheel', feedWheel as EventListener, { passive: false })
    targets.push(target)

    return () => unobserve(target)
  }

  const unobserve = (target: EventTarget) => {
    target.removeEventListener('wheel', feedWheel as EventListener)
    targets = targets.filter((t) => t !== target)
  }

  // stops watching all of its target elements for visibility changes.
  const disconnect = () => {
    targets.forEach(unobserve)
  }

  const subscribe = (callback: SubscribeFn): Unsubscribe => {
    subscriptions.push(callback)
    // return bound unsubscribe
    return () => unsubscribe(callback)
  }

  const unsubscribe = (callback: SubscribeFn) => {
    if (!callback) throw new Error('please pass the callback which was used to subscribe')
    subscriptions = subscriptions.filter((s) => s !== callback)
  }

  const publish = (type: WheelPhase, additionalData?: Partial<PhaseData>) => {
    const data: PhaseData = {
      type,
      isEndingSoon: willEndSoon(),
      isMomentum: state.isMomentum,
      axisMovement: state.axisMovement,
      axisVelocity: state.axisVelocity,
      event: currentEvent,
      ...additionalData,
    }
    subscriptions.forEach((fn) => fn(type, data))
  }

  const feedWheel = (wheelEvents: WheelEventData | WheelEventData[]) => {
    if (Array.isArray(wheelEvents)) {
      wheelEvents.forEach((wheelEvent) => processWheelEventData(wheelEvent))
    } else {
      processWheelEventData(wheelEvents)
    }
  }

  const shouldPreventDefault = (e: WheelEventData) => {
    const { preventWheelAction } = options
    const { deltaX, deltaY } = e

    switch (preventWheelAction) {
      case 'all':
        return true
      case 'x':
        return Math.abs(deltaX) >= Math.abs(deltaY)
      case 'y':
        return Math.abs(deltaY) >= Math.abs(deltaX)
    }

    isDev && console.warn('unsupported preventWheelAction value: ' + preventWheelAction, 'warn')
  }

  const clampDelta = (delta: number) => {
    return Math.min(DELTA_MAX_ABS, Math.max(-DELTA_MAX_ABS, delta))
  }

  const processWheelEventData = (wheelEvent: WheelEventData) => {
    const normalizedWheel = normalizeWheel(wheelEvent)

    if (wheelEvent.preventDefault && shouldPreventDefault(wheelEvent)) {
      wheelEvent.preventDefault()
    }

    if (!state.isStarted) {
      start()
    }

    const currentDelta = clampDelta(
      Math.abs(normalizedWheel.deltaY) > Math.abs(normalizedWheel.deltaX)
        ? normalizedWheel.deltaY
        : normalizedWheel.deltaX
    )
    const currentAbsDelta = Math.abs(currentDelta)

    if (state.isMomentum && currentAbsDelta > state.lastAbsDelta) {
      end()
      start()
    }

    currentEvent = wheelEvent

    state.axisMovement = state.axisMovement.map(
      (prevDelta, i) => prevDelta + clampDelta(normalizedWheel[deltaProp[axes[i]]])
    )

    state.lastAbsDelta = currentAbsDelta

    state.scrollPointsToMerge.push({
      currentAbsDelta: currentAbsDelta,
      axisDeltaUnclampt: [normalizedWheel.deltaX, normalizedWheel.deltaY],
      timestamp: wheelEvent.timeStamp,
    })

    if (state.scrollPointsToMerge.length === WHEELEVENTS_TO_MERGE) {
      const { scrollPointsToMerge } = state
      const mergedScrollPoint: ScrollPoint = {
        currentAbsDelta: average(scrollPointsToMerge.map((b) => b.currentAbsDelta)),
        axisDeltaUnclampt: scrollPointsToMerge.reduce(
          ([sumX, sumY], { axisDeltaUnclampt: [x, y] }) => [sumX + x, sumY + y],
          [0, 0]
        ),
        timestamp: average(scrollPointsToMerge.map((b) => b.timestamp)),
      }

      state.scrollPoints.push(mergedScrollPoint)

      // only update velocity after a merged scrollpoint was generated
      updateVelocity()

      if (!state.isMomentum) {
        detectMomentum()
      }

      // reset merge array
      state.scrollPointsToMerge = []
    }

    if (!state.scrollPoints.length) {
      updateStartVelocity()
    }

    // publish start after velocity etc. have been updated
    if (!state.isStartPublished) {
      publish(WheelPhase.ANY_WHEEL_START)
      publish(WheelPhase.WHEEL_START)
      state.isStartPublished = true
    }

    publish(WheelPhase.ANY_WHEEL)
    publish(state.isMomentum ? WheelPhase.MOMENTUM_WHEEL : WheelPhase.WHEEL)

    // calc debounced end function, to recognize end of wheel event stream
    willEnd()
  }

  const updateStartVelocity = () => {
    const latestScrollPoint = state.scrollPointsToMerge[state.scrollPointsToMerge.length - 1]
    state.axisVelocity = latestScrollPoint.axisDeltaUnclampt.map((d) => d / state.willEndTimeout)
  }

  const updateVelocity = () => {
    // need to have two recent points to calc velocity
    const [pA, pB] = state.scrollPoints.slice(-2)

    if (!pA || !pB) {
      return
    }

    // time delta
    const deltaTime = pB.timestamp - pA.timestamp

    if (deltaTime <= 0) {
      isDev && console.warn('invalid deltaTime')
      return
    }

    // calc the velocity per axes
    const velocity = pB.axisDeltaUnclampt.map((d) => d / deltaTime)

    // calc the acceleration factor per axis
    const accelerationFactor = velocity.map((v, i) => {
      return v / (state.axisVelocity[i] || 1)
    })

    state.axisVelocity = velocity
    state.accelerationFactors.push(accelerationFactor)
    updateWillEndTimeout(deltaTime)
  }

  const updateWillEndTimeout = (deltaTime: number) => {
    // use current time between events rounded up and increased by a bit as timeout
    let newTimeout = Math.ceil(deltaTime / 10) * 10 * 1.2

    // double the timeout, when momentum was not detected yet
    if (!state.isMomentum) {
      newTimeout = Math.max(100, newTimeout * 2)
    }

    state.willEndTimeout = Math.min(1000, Math.round(newTimeout))
  }

  const accelerationFactorInMomentumRange = (accFactor: number) => {
    // when main axis is the the other one and there is no movement/change on the current one
    if (accFactor === 0) return true
    return accFactor <= ACC_FACTOR_MAX && accFactor >= ACC_FACTOR_MIN
  }

  const detectMomentum = () => {
    if (state.accelerationFactors.length < WHEELEVENTS_TO_ANALAZE) {
      return state.isMomentum
    }

    const recentAccelerationFactors = state.accelerationFactors.slice(WHEELEVENTS_TO_ANALAZE * -1)

    // check recent acceleration / deceleration factors
    const detectedMomentum = recentAccelerationFactors.reduce((mightBeMomentum, accFac) => {
      // all recent need to match, if any did not match -> short circuit
      if (!mightBeMomentum) return false
      // when both axis decelerate exactly in the same rate it is very likely caused by momentum
      const sameAccFac = !!accFac.reduce((f1, f2) => (f1 && f1 < 1 && f1 === f2 ? 1 : 0))
      // check if acceleration factor is within momentum range
      const bothAreInRangeOrZero = accFac.filter(accelerationFactorInMomentumRange).length === accFac.length
      // one the requirements must be fulfilled
      return sameAccFac || bothAreInRangeOrZero
    }, true)

    // only keep the most recent events
    state.accelerationFactors = recentAccelerationFactors

    if (detectedMomentum && !state.isMomentum) {
      publish(WheelPhase.WHEEL_END)
      state.isMomentum = true
      publish(WheelPhase.MOMENTUM_WHEEL_START)
    }

    return state.isMomentum
  }

  const start = () => {
    state = createWheelAnalyzerState()
    state.isStarted = true
  }

  const willEnd = (() => {
    let willEndId: NodeJS.Timeout
    return () => {
      clearTimeout(willEndId)
      willEndId = setTimeout(end, state.willEndTimeout)
    }
  })()

  const end = () => {
    if (!state.isStarted) return

    if (state.isMomentum) {
      if (!willEndSoon()) {
        publish(WheelPhase.MOMENTUM_WHEEL_CANCEL)
      } else {
        publish(WheelPhase.MOMENTUM_WHEEL_END)
      }
    } else {
      // in case of momentum, this event was already triggered when the momentum was detected so we do not trigger it here
      publish(WheelPhase.WHEEL_END)
    }

    publish(WheelPhase.ANY_WHEEL_END)

    state.isMomentum = false
    state.isStarted = false
  }

  const willEndSoon = () => {
    const absDeltas = state.scrollPoints
      .slice(SOON_ENDING_WHEEL_COUNT * -1)
      .map(({ currentAbsDelta }) => currentAbsDelta)

    const absDeltaAverage = absDeltas.reduce((a, b) => a + b, 0) / absDeltas.length
    return absDeltaAverage <= SOON_ENDING_THRESHOLD
  }

  return Object.freeze({
    observe,
    unobserve,
    disconnect,
    subscribe,
    unsubscribe,
    feedWheel,
  })
}
