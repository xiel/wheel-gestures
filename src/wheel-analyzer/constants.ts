import { Axis, DeltaProp } from './wheel-analyzer-types'

export const SOON_ENDING_THRESHOLD = 1.4
export const ACC_FACTOR_MIN = 0.6
export const ACC_FACTOR_MAX = 0.96
export const DELTA_MAX_ABS = 150

/**
 * the timeout is automatically adjusted during a gesture
 * the initial timeout period is pretty long, so even old mouses, which emit wheel events less often, can produce a continuous gesture
 */
export const WILL_END_TIMEOUT_DEFAULT = 400

export const axes: [Axis, Axis] = ['x', 'y']
export const deltaProp: Record<Axis, DeltaProp> = {
  x: 'deltaX',
  y: 'deltaY',
}

export const isDev = process.env.NODE_ENV !== 'production'
export const WHEELEVENTS_TO_MERGE = 2
export const WHEELEVENTS_TO_ANALAZE = 5

export const SOON_ENDING_WHEEL_COUNT = 3
