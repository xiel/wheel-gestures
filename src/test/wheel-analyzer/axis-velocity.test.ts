import { subscribeAndFeedWheelEvents } from '../helper/recordPhases'
import { PhaseData, WheelEventData } from '../../wheel-analyzer.types'

interface GenerateEventsProps {
  deltaTotal: number[]
  durationMs: number
  eventEveryMs?: number
  deltaMode?: number
}

function generateEvents({ deltaTotal, durationMs, eventEveryMs = 1000 / 60, deltaMode = 0 }: GenerateEventsProps) {
  const wheelEvents: WheelEventData[] = []
  const [deltaX, deltaY, deltaZ] = deltaTotal.map((d) => (d / durationMs) * eventEveryMs)

  let timeStamp = 0
  while (timeStamp < durationMs && timeStamp + eventEveryMs < durationMs) {
    timeStamp += eventEveryMs

    wheelEvents.push({
      deltaX,
      deltaY,
      deltaZ,
      deltaMode,
      timeStamp,
    })
  }

  return { wheelEvents }
}

function calcAverageVelocity(phases: PhaseData[]) {
  return phases.reduce(
    (averageAcc, { axisVelocity }) => {
      averageAcc[0] = (averageAcc[0] + axisVelocity[0]) / 2
      averageAcc[1] = (averageAcc[1] + axisVelocity[1]) / 2
      return averageAcc
    },
    [0, 0]
  )
}

describe('velocity', () => {
  it('should have velocity in first (and only) event - x', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents({
      wheelEvents: [{ deltaX: 0, deltaY: 20, deltaMode: 0 }],
    })
    const [xVelo, yVelo] = allPhaseData[1].axisVelocity

    // velocity in first event is based on the initial timeout between two events
    expect(xVelo).toEqual(0)
    expect(yVelo).toEqual(20 / 400)
    expect(allPhaseData).toMatchSnapshot()
  })

  it('check average velocity', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(generateEvents({ deltaTotal: [0, 500], durationMs: 1000 }))
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(0.5)
  })

  it('check average velocity 2', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(generateEvents({ deltaTotal: [0, 1000], durationMs: 1000 }))
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(1)
  })

  it('check average velocity 3', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(
      generateEvents({ deltaTotal: [0, 1000], durationMs: 1000, eventEveryMs: 1000 / 30 })
    )
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(1)
  })
})
