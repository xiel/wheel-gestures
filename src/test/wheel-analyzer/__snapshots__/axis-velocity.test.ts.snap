// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`velocity should have velocity in first (and only) event - x 1`] = `
Array [
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "ANY_WHEEL_START",
    "willEndSoon": false,
  },
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "WHEEL_START",
    "willEndSoon": false,
  },
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "ANY_WHEEL",
    "willEndSoon": false,
  },
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "WHEEL",
    "willEndSoon": false,
  },
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "WHEEL_END",
    "willEndSoon": false,
  },
  Object {
    "axisMovement": Array [
      0,
      20,
    ],
    "axisVelocity": Array [
      0,
      0.05,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isMomentum": false,
    "type": "ANY_WHEEL_END",
    "willEndSoon": false,
  },
]
`;
